### PR TITLE
Fix ANYCUBIC_LCD_CHIRON compilation

### LINKED ISSUE #20805: Anycubic Chiron w/ stock LCD w/ SKR 1.4 Turbo compile fails
### Bug Description

I have an Anycubic Chiron which I am upgrading to an SKR 1.4 Turbo and a combination of 2130 and 2208 drivers.  I want to use the original screen on the Chiron.  I have Marlin compiling all the way to "Linking...firmware.elf", then ld.exe gives me a bunch of errors about "undefined reference to 'Anycubic::Chiron'".  I cannot for the life of me figure out what it's talking about, since all reference appear to be in the header files, as needed, and the respective .o files are in the folder it specifies.  
I have never used Marlin (always used Klipper on my other machine), but I would very much like to get this working with the original screen and the SKR.  Any help would be appreciated.  
I am using Visual Studio Code with PlatformIO (which I've never used before, either) so it is possible that I have an environment issue, though I don't believe that's it.  
Thanks in advance!

### Configuration Files
[Marlin.zip](https://github.com/MarlinFirmware/Marlin/files/5827333/Marlin.zip)
[platformio.zip](https://github.com/MarlinFirmware/Marlin/files/5827350/platformio.zip)

I had to add the following pin definition to pins_BTT_SKR_V1_4.h:
#ifndef Z_MAX_PIN
     #define Z_MAX_PIN                        P1_00  // PWRDET
#endif

and the following pin definition to pins_BTT_SKR_V1_4_TURBO.h (no reason I did it in two separate files, just worked out that way.)
#ifndef BEEPER_PIN
    #define BEEPER_PIN P1_30
#endif

### Steps to Reproduce

1. Make pin definitions listed above
2. Copy configuration[_adv].h files
3. Compile with PlatformIO

**Expected behavior:**

Successful Compile

**Actual behavior:**

Everything compiles fine until "Linking .pio\build\LPC1769\firmware.elf" when I get the following errors and fails:

c:/users/brian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: .pio/build/LPC1769/src/src/lcd/extui/anycubic_chiron_lcd.cpp.o: in function `ExtUI::onStartup()':
C:\Users\brian\Downloads\Marlin-bugfix-2.0.x (1)\Marlin-bugfix-2.0.x/Marlin\src\lcd\extui/anycubic_chiron_lcd.cpp:40: undefined reference to `Anycubic::Chiron'
c:/users/brian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: .pio/build/LPC1769/src/src/lcd/extui/anycubic_chiron_lcd.cpp.o: in function `ExtUI::onIdle()':
C:\Users\brian\Downloads\Marlin-bugfix-2.0.x (1)\Marlin-bugfix-2.0.x/Marlin\src\lcd\extui/anycubic_chiron_lcd.cpp:42: undefined reference to `Anycubic::Chiron'
c:/users/brian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: .pio/build/LPC1769/src/src/lcd/extui/anycubic_chiron_lcd.cpp.o: in function `ExtUI::onPrinterKilled(char const*, char const*)':
C:\Users\brian\Downloads\Marlin-bugfix-2.0.x (1)\Marlin-bugfix-2.0.x/Marlin\src\lcd\extui/anycubic_chiron_lcd.cpp:45: undefined reference to `Anycubic::Chiron'
c:/users/brian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: .pio/build/LPC1769/src/src/lcd/extui/anycubic_chiron_lcd.cpp.o: in function `ExtUI::onMediaInserted()':
C:\Users\brian\Downloads\Marlin-bugfix-2.0.x (1)\Marlin-bugfix-2.0.x/Marlin\src\lcd\extui/anycubic_chiron_lcd.cpp:48: undefined reference to `Anycubic::Chiron'
c:/users/brian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: .pio/build/LPC1769/src/src/lcd/extui/anycubic_chiron_lcd.cpp.o: in function `ExtUI::onMediaError()':
C:\Users\brian\Downloads\Marlin-bugfix-2.0.x (1)\Marlin-bugfix-2.0.x/Marlin\src\lcd\extui/anycubic_chiron_lcd.cpp:49: undefined reference to `Anycubic::Chiron'
c:/users/brian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: .pio/build/LPC1769/src/src/lcd/extui/anycubic_chiron_lcd.cpp.o:C:\Users\brian\Downloads\Marlin-bugfix-2.0.x (1)\Marlin-bugfix-2.0.x/Marlin\src\lcd\extui/anycubic_chiron_lcd.cpp:50: more undefined references to `Anycubic::Chiron' follow
c:/users/brian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: .pio/build/LPC1769/src/src/lcd/extui/lib/anycubic_chiron/chiron_tft.cpp.o: in function `Anycubic::ChironTFT::SendFileList(signed char)':
C:\Users\brian\Downloads\Marlin-bugfix-2.0.x (1)\Marlin-bugfix-2.0.x/Marlin\src\lcd\extui\lib\anycubic_chiron/chiron_tft.cpp:398: undefined reference to `Anycubic::filenavigator'
c:/users/brian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: .pio/build/LPC1769/src/src/lcd/extui/lib/anycubic_chiron/chiron_tft.cpp.o: in function `Anycubic::ChironTFT::SelectFile()':
C:\Users\brian\Downloads\Marlin-bugfix-2.0.x (1)\Marlin-bugfix-2.0.x/Marlin\src\lcd\extui\lib\anycubic_chiron/chiron_tft.cpp:420: undefined reference to `Anycubic::filenavigator'
c:/users/brian/.platformio/packages/toolchain-gccarmnoneeabi/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: .pio/build/LPC1769/src/src/lcd/extui/lib/anycubic_chiron/chiron_tft.cpp.o: in function `Anycubic::ChironTFT::PanelAction(unsigned char)':
C:\Users\brian\Downloads\Marlin-bugfix-2.0.x (1)\Marlin-bugfix-2.0.x/Marlin\src\lcd\extui\lib\anycubic_chiron/chiron_tft.cpp:692: undefined reference to `Anycubic::filenavigator'
collect2.exe: error: ld returned 1 exit status
*** [.pio\build\LPC1769\firmware.elf] Error 1
================================================================================ [FAILED] Took 167.26 seconds ================================================================================
Environment    Status    Duration
-------------  --------  ------------
LPC1769        FAILED    00:02:47.261
=========================================================================== 1 failed, 0 succeeded in 00:02:47.261 =========================================================================== 
The terminal process "C:\Users\brian\.platformio\penv\Scripts\pio.exe 'run'" terminated with exit code: 1.

#### Additional Information

Marlin Error

### PR DESCRIPTION
**Pull request by sjasonsmith**
### Description

This fixes several compilation issues impacting the AnyCubic Chiron display:

1. Missing global variable instantiations
2. Allow const string pointers to be passed to `injectCommands`.
3. Two issues with power-loss recovery (I will comment on these in the code)
    - Misleading indentation, and _probably_ incorrect `if` behavior.
    - Two subsequent calls to `injectCommands_P`. This is invalid, because the `_P` variant stores a single pointer. This guarantees that only the second call will be retained.

### Requirements

SKR 1.4 Turbo with AnyCubic Chiron display

### Benefits

Allows compilation, although I am unable to test this myself.

### Configurations

These are the same as in the reported bug, but I have included the pin customizations in the Configuration file.
[Bug_20805.zip](https://github.com/MarlinFirmware/Marlin/files/5828016/Bug_20805.zip)

### Related Issues

#20805

### CODE REVIEW

File: Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft.cpp [hunk 1]
```diff
@@ -574,10 +576,11 @@ namespace Anycubic {
...
 
       case 15:   // A15 Resuming from outage
-        if (printer_state == AC_printer_resuming_from_power_outage)
+        if (printer_state == AC_printer_resuming_from_power_outage) {
           // Need to home here to restore the Z position
-          injectCommands_P(AC_cmnd_power_loss_recovery);
-          injectCommands_P(PSTR("M1000"));  // home and start recovery
+          injectCommands(AC_cmnd_power_loss_recovery);
+          injectCommands("M1000");  // home and start recovery
+        }
```
Comment (sjasonsmith) [inline Marlin/src/lcd/extui/lib/anycubic_chiron/chiron_tft.cpp:583]: It is unclear from the existing indentation whether or not the `M1000` is intended to be conditional. I added braces _presuming_ that both were supposed to be inside the `if` statement.

If this is incorrect, the indentation will need to be fixed and the braces removed.

In the existing code `M1000` would have always replaced the earlier inserted command, since injectCommands_P stores a single string pointer. Switching to `injectCommands` should allow two commands to be injected, although I am not an expert in the proper usage of these commands. I also don't know if this display is ever used on AVR, and whether passing a PSTR to `injectCommands` might be a problem.

**Comment by sjasonsmith** [comment]
Hopefully @BrianP310 will test this out, since they are his configs.

**Comment by BrianP310** [comment]
[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/5831002/Configuration.zip)

@sjasonsmith Thanks, that definitely allows it to compile now.  I had to change my Configuration.h because the TFT connector on the SKR 1.4T is Serial 0, so has to set 
```cpp
#define LCD_SERIAL_PORT 0.  
```
This makes the following changes necessary:
```cpp
#define SERIAL_PORT 1
#define SERIAL_PORT_2 -1
```
Also, annoyingly, the stock Chiron LCD has the TFT connector wired in a non-standard way.  So to make it work with the SKR 1.4T, you have to swap the pins around as follows (Luckily, if your colors are different, the pins definitions are silk screened on the back of the LCD circuit board, so just match up the colors, as necessary.)
```
|--------------------------|
|    G    Bk   Y   R   Bl  |   G=RST, Bk=Tx (Goes to Rx on SKR), Y=Rx (Goes to Tx), R=Vss (Gnd), Bl=5V
|------|----------|--------|
```
So, to summarize, in total to fix this bug, we need to merge your code changes, add the two pin definitions:
```cpp
#ifndef BEEPER_PIN
    #define BEEPER_PIN P1_30
#endif

//
// Second Z-Stop Pin
//
#ifndef Z_MAX_PIN
  #define Z_MAX_PIN                        P1_00  // PWRDET
#endif
```
and rewire the TFT connector.  Seems this was definitely a bug, but it is working now.

**Comment by thinkyhead** [comment]
> So, to summarize, in total to fix this bug, we need to merge your code changes, add the two pin definitions:

It seems that `#define BEEPER_PIN EXPA1_10_PIN` actually does belong under `ANET_FULL_GRAPHICS_LCD` in the BTT SKR 1.4 pins file as a standard thing. The pin there is labeled as "beeper."

What is the `Z_MAX_PIN` used for if it's not being used as an endstop?

**Comment by thisiskeithb** [comment]
> What is the Z_MAX_PIN used for if it's not being used as an endstop?

That's the thing with the [SKR 1.4](https://github.com/bigtreetech/BIGTREETECH-SKR-V1.3/blob/master/BTT%20SKR%20V1.4/Hardware/BTT%20SKR%20V1.4PIN.pdf). There's no Z+ pin, just a Z stop pin. The silkscreen says `PWRDET` where Z+ was located previously on the SKR 1.3:
image.

As you know, many pins can be remapped and users preferred the default [SKR 1.3](https://github.com/bigtreetech/BIGTREETECH-SKR-V1.3/blob/master/BTT%20SKR%20V1.3/hardware/SKR-V1.3-PIN.pdf) pins when upgrading, so I think that's why much of that bleeds over into the SKR 1.4 pins.

**Comment by BrianP310** [comment]
It is being used as the second endstop in the dual z stepper setup, however it isn't defined when you go to compile.

**Comment by thinkyhead** [comment]
The Anycubic Chiron uses the `OUTAGECON_PIN` to send voltage which is detected by the `POWER_LOSS_PIN` (pin 79 / Port I, bit 7). This pin is only defined in the Trigorilla configuration. It's not clear what these pins will be on the SKR 1.4.

**Comment by BrianP310** [comment]
I don't mean for this to sound as snarky as it does, but I don't understand what that has to do with with the SKR 1.4T LCD screen problem. What am I missing? The above mentioned fix makes everything work.

**Comment by thinkyhead** [comment]
> It is being used as the second endstop in the dual z stepper setup…

Maybe we can add special handling of endstops in some of the pins files when the dual endstop options are enabled. The current handling only looks at sensorless homing is in use, and makes the "other endstop" available based on the primary homing direction. When dual endstops are enabled the meaning of "the other endstop" is not going to be quite the same.

**Comment by thinkyhead** [comment]
> I don't mean for this to sound as snarky as it does, but I don't understand what that has to do with with the SKR 1.4T LCD screen problem. What am I missing? The above mentioned fix makes everything work.

Since you're using a whole different MOTHERBOARD, I've been putting together an example configuration for it. So I need to know how to fill out the whole thing.

**Comment by BrianP310** [comment]
I would agree with that. I'm not a programmer (at least not in C) but that is how I was able to get it to work. How you guys think it should be "properly" implemented is probably way better than what I came up with. I just appreciate the help. I thought I was was losing my mind when I couldn't get it working, so I'm quite happy to have found an actual bug. Now if only I could get the stock bed leveling sensor to quit crashing 10mm past the bed, it'll be perfect!

**Comment by BrianP310** [comment]
I went through and did a proper wiring job, so not sure if this helps you, but here it is all wired up.

**Comment by stef-ladefense** [comment]
> Port I, bit 7).

hi,
it's more like J7 isn't it?
besides since J7 is mapped on DIO74 (ref fastio_1280.h, hard 79 pin)
can me show me on a photo of the trigorilla (8bit) where is D74?

**Comment by cofhal** [comment]
Could this configuration file be used on the Anycubic MEGA S, with the original TFT with the SKR 1.4?

**Comment by thomin07** [comment]
Hello everyone, I would like to come here because I have a chrion which has just broken down, I have just received an SKR 1.4 turbo, do you have a wiring diagram knowing that I am keeping the original PCBs. If you also have links to the firmware, thank you in advance

**Comment by JC-GOMEZ-15** [comment]
hey I´m trying to upgrade my anycubic chiron with the Bigtreetech SKR 1.4 mainboard, do you have any firmware for me to use?
